src/Plugin/PasswordConstraint/MinimumCharacters.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Drupal\password_enhancements\Plugin\PasswordConstraint;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\PluginBase;
+use Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface;
+
+/**
+ * Defines a common class for minimum character requirement.
+ */
+abstract class MinimumCharacters extends PluginBase implements PasswordConstraintPluginInterface {
+
+  /**
+   * The generated error message.
+   *
+   * @var string
+   */
+  protected $errorMessage;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->errorMessage = NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getErrorMessage(): ?string {
+    return $this->errorMessage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getInitialDescription(): string {
+    if ($this->configuration['minimum_characters'] > 1) {
+      return $this->configuration['descriptionPlural'];
+    }
+    else {
+      return $this->configuration['descriptionSingular'];
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(): array {
+    $form = [];
+
+    $form['minimum_characters'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Minimum characters'),
+      '#required' => TRUE,
+      '#min' => 1,
+      '#default_value' => !empty($this->configuration['minimum_characters']) ? $this->configuration['minimum_characters'] : 1,
+    ];
+
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsValidate(array &$form, FormStateInterface $form_state): void {
+    if ($form_state->getValue(['settings', 'minimum_characters']) < 1) {
+      $form_state->setError($form['settings']['minimum_characters'], $this->t('The %title field must be a non-zero, positive number.', [
+        '%title' => $form['settings']['minimum_characters']['#title'],
+      ]));
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsSubmit(array &$form, FormStateInterface $form_state): void {
+    // Nothing to do here.
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate(string $value, bool $custom_error_message = FALSE): bool {
```

Comment:
The validate function validates the password against a defined constraint, so it must return true or false, so if in case if it fails it should set an error message, but I did not wanted to return the error message here because that's expensive and probably not trivial to check.